upstream(indexer): pruner updates watermarks lower bound

tomxey opened the pull request:
# Description of change

Original description:

> With the committer writing upper bounds, the pruner can now read from watermarks and determine whether the lower bounds need to be updated. Pruner does this by spawning a separate task, without touching the extant pruning logic (so all things are as is.) It will ignore any entries from watermarks that do not correspond to a PrunableTable variant.

## Links to any relevant issues

fixes https://github.com/iotaledger/iota/issues/8731

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)

### Infrastructure QA (only required for crates that are maintained by @iotaledger/infrastructure)

- [ ] Synchronization of the indexer from genesis for a network including migration objects.
- [ ] Restart of indexer synchronization locally without resetting the database.
- [ ] Restart of indexer synchronization on a production-like database.
- [ ] Deployment of services using Docker.
- [ ] Verification of API backward compatibility.

tomxey committed:
upstream(indexer): pruner updates watermarks lower bound

kodemartin commented on crates/iota-indexer/src/models/watermarks.rs:
```diff
@@ -44,7 +47,7 @@ pub struct StoredWatermark {
     /// Column used by the pruner to track its true progress. Data at and below
     /// this watermark can be immediately pruned.
-    pub pruner_lo: Option<i64>,
+    pub pruner_hi_inclusive: Option<i64>,
```
What purpose lies behind this change. I don't see this being used, and it seems incosistent with its documentation.

tomxey replied:
For pruner, this is the highest value that can be immediately deleted, and it's inclusive, so I would say the name makes sense for now. It's not used yet though so it still can change potentially.

kodemartin commented on crates/iota-indexer/src/pruning/pruner.rs:
```diff
@@ -179,3 +222,58 @@ impl Pruner {
+    retention_policies: HashMap<PrunableTable, u64>,
+    cancel: CancellationToken,
+) -> IndexerResult<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(5));
```
I would suggest a `const` instead of hardcoding the duration here.

tomxey replied:
Done in 63fb502058b765876b2e841a8f6607e70da8b76c

kodemartin commented on crates/iota-indexer/src/pruning/pruner.rs:
```diff
@@ -179,3 +222,58 @@ impl Pruner {
+        if cancel.is_cancelled() {
+            info!("Pruner watermark lower bound update task cancelled.");
+            return Ok(());
+        }
+        let Some(prunable_table) = watermark.entity() else {
+            continue;
+        };
+        let Some(epochs_to_keep) = retention_policies.get(&prunable_table) else {
+            error!("no retention policy found for prunable table {prunable_table}");
+            continue;
+        };
+        if let Some(new_epoch_lo) = watermark.new_epoch_lo(*epochs_to_keep) {
+            lower_bound_updates.push((prunable_table, new_epoch_lo));
+        };
+    }
```
I think it is plausible to let the loop finish gracefully updating all watermarks, making the update atomic.

Cancellation can be checked before the loop.

tomxey replied:
The update is atomic, as it happens once after the loop. The loop is only used to collect the values that need to be updated, so I think it's fine if we return in the middle of it.

sergiupopescu199 replied:
I do agree with @kodemartin, 

Looking at the `CancellationToken` [implementation](https://docs.rs/tokio-util/latest/src/tokio_util/sync/cancellation_token.rs.html#206) calling frequently the `is_cancelled` method could add some overhead, inside it has a list of nodes, looking at its [implementation](https://docs.rs/tokio-util/latest/src/tokio_util/sync/cancellation_token/tree_node.rs.html#83). 

It does lock the Mutex every time we call this method, unless we have tens of thousands of records, the performance cost should be negligible, I think having the cancellation token on the `tokio::select` could be better, makes also the flow clearer, we either exit on cancellation token or we do update the lower bounds.

kodemartin replied:
The db operation at the end is indeed atomic. My point is that once we start preparing the watermarks for update, I think graceful shutdown implies letting the preparation to finish and update the data.

tomxey replied:
Moved the check before the loop
469bb4131fa605df6d110d67e820f0921eb00bd6

kodemartin reviewed (commented):
Looks good with some minor comments/suggestions.

tomxey committed:
refactor(indexer): Store duration as const

sergiupopescu199 commented on crates/iota-indexer/src/pruning/pruner.rs:
```diff
@@ -73,6 +79,36 @@ pub enum PrunableTable {
     PrunerCpWatermark,
 }
 
+impl PrunableTable {
+    pub fn select_reader_lo(&self, cp: u64, tx: u64) -> u64 {
+        match self {
+            PrunableTable::ObjectsHistory => cp,
+            PrunableTable::Transactions => tx,
+            PrunableTable::Events => tx,
+            PrunableTable::EventEmitPackage => tx,
+            PrunableTable::EventEmitModule => tx,
+            PrunableTable::EventSenders => tx,
+            PrunableTable::EventStructInstantiation => tx,
+            PrunableTable::EventStructModule => tx,
+            PrunableTable::EventStructName => tx,
+            PrunableTable::EventStructPackage => tx,
+            PrunableTable::TxAffectedAddresses => tx,
+            PrunableTable::TxAffectedObjects => tx,
+            PrunableTable::TxCallsPkg => tx,
+            PrunableTable::TxCallsMod => tx,
+            PrunableTable::TxCallsFun => tx,
+            PrunableTable::TxChangedObjects => tx,
+            PrunableTable::TxDigests => tx,
+            PrunableTable::TxInputObjects => tx,
+            PrunableTable::TxKinds => tx,
+            PrunableTable::TxRecipients => tx,
+            PrunableTable::TxSenders => tx,
+            PrunableTable::Checkpoints => cp,
+            PrunableTable::PrunerCpWatermark => cp,
+        }
+    }
```
This version could improve a bit the readability.

```suggestion
pub fn select_reader_lo(&self, cp: u64, tx: u64) -> u64 {
    match self {
        PrunableTable::ObjectsHistory
        | PrunableTable::Checkpoints
        | PrunableTable::PrunerCpWatermark => cp,

        PrunableTable::Transactions
        | PrunableTable::Events
        | PrunableTable::EventEmitPackage
        | PrunableTable::EventEmitModule
        | PrunableTable::EventSenders
        | PrunableTable::EventStructInstantiation
        | PrunableTable::EventStructModule
        | PrunableTable::EventStructName
        | PrunableTable::EventStructPackage
        | PrunableTable::TxAffectedAddresses
        | PrunableTable::TxAffectedObjects
        | PrunableTable::TxCallsPkg
        | PrunableTable::TxCallsMod
        | PrunableTable::TxCallsFun
        | PrunableTable::TxChangedObjects
        | PrunableTable::TxDigests
        | PrunableTable::TxInputObjects
        | PrunableTable::TxKinds
        | PrunableTable::TxRecipients
        | PrunableTable::TxSenders => tx,
    }
}
```

tomxey replied:
Done in b5d97cdd563be00a94e8a449186821375b70fa97

sergiupopescu199 reviewed (approved):
LGTM 🌸

tomxey committed:
refactor(indexer): Make the mapping more readable

Co-authored-by: Sergiu Popescu <44298302+sergiupopescu199@users.noreply.github.com>

tomxey committed:
feat(indexer): Check cancellation token once before the loop

tomxey committed:
upstream(indexer): align watermarks table schema in live indexer to alt indexer (#9124)

# Description of change

Original description:


> Since watermarks table isn't being written to yet, modify the db
schema to match alt-indexer. The changes are to rename entity ->
pipeline, tx_hi_inclusive -> tx_hi, and pruner_hi_inclusive -> pruner_hi
and make it a non-null column. This works out nicely for graphql, since
the transactions query implementations expect a half-open interval. Also
simplifies pruner logic, since it can write the reader_lo as pruner_hi
after delay, and table pruners will delete between [table_data,
pruner_hi).


## Links to any relevant issues

fixes https://github.com/iotaledger/iota/issues/9122

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)

### Infrastructure QA (only required for crates that are maintained by
@iotaledger/infrastructure)

- [ ] Synchronization of the indexer from genesis for a network
including migration objects.
- [ ] Restart of indexer synchronization locally without resetting the
database.
- [ ] Restart of indexer synchronization on a production-like database.
- [ ] Deployment of services using Docker.
- [ ] Verification of API backward compatibility.
